core/src/main/java/com/linecorp/armeria/internal/server/annotation/AnnotatedService.java

```diff
@@ -455,28 +454,27 @@ private HttpResponse convertResponse(ServiceRequestContext ctx, @Nullable HttpHe
         }
     }
 
-    private static ResponseHeadersBuilder addNegotiatedResponseMediaType(ServiceRequestContext ctx,
-                                                                         HttpHeaders headers) {
-
+    private static ResponseHeaders addNegotiatedResponseMediaType(ServiceRequestContext ctx,
+                                                                  ResponseHeadersBuilder headers) {
+        // :status header is guaranteed to be set.
+        if (headers.status().isContentAlwaysEmpty()) {
+            return headers.build();
+        }
         final MediaType negotiatedResponseMediaType = ctx.negotiatedResponseMediaType();
         if (negotiatedResponseMediaType == null || headers.contentType() != null) {
             // Do not overwrite 'content-type'.
-            return ResponseHeaders.builder()
-                                  .add(headers);
+            return headers.build();
         }
-
-        return ResponseHeaders.builder()
-                              .add(headers)
-                              .contentType(negotiatedResponseMediaType);
+        return headers.contentType(negotiatedResponseMediaType).build();
     }
 
-    private ResponseHeaders setHttpStatus(ResponseHeadersBuilder headers) {
+    private ResponseHeadersBuilder setHttpStatus(HttpHeaders headers) {
+        final ResponseHeadersBuilder builder = ResponseHeaders.builder().add(headers);
```

Comment:
`builder.add()` is an expensive operation. It would be better to avoid calling it if `headers` is an instance of `ResponseHeaders`
```suggestion
        final ResponseHeadersBuilder builder;
        if (headers instanceof ResponseHeaders) {
            header = ((ResponseHeaders) headers).toBuilder();
        } else {
            builder = ResponseHeaders.builder().add(headers);
        }
```

Reply:
Refactored `convertResponse` method taking your comments into account.
Sorry that I updated the code too late.